packages/lib/storage/src/new/MessageStorage.ts

```diff
@@ -0,0 +1,399 @@
+        sizeLimit: number,
+        key: string,
+        children: Node[] | Envelop[],
+    ) {
+        const instance = new Conversation(db, enc, sizeLimit, key, children);
+        await instance.save();
+        return instance;
+    }
+    protected override serialize() {
+        //Conversations are serialited according the the spec
+        const mapped = this.getLeafs<Node>().map((c) => ({
+            id: c.key,
+            timestamp: c.getLeafs<Envelop>()[0].message.metadata.timestamp,
+        }));
+
+        return JSON.stringify(mapped);
+    }
+
+    public static computeKey(rootKey: string, to: string) {
+        return sha256(rootKey + to);
+    }
+
+    public async add(message: Envelop) {
+        const conversationIsEmpty = this.getLeafs<Node>().length === 0;
+        // the first chunk has to be created
+        if (conversationIsEmpty) {
+            await this.createNewChunkAndAddMessage(message);
+            return;
+        }
+
+        //get the latest chunk
+        const latestChunk = this.getLeafs<Chunk>()[this.getLeafs().length - 1];
+        //Fetch all envelops of the chunk to determine if the chunk is full
+        await latestChunk.fetch();
+
+        //Check if the chunk is full if the message would be added
+        const chunkIsFull = !latestChunk.hasSpace(message);
+        //If the chunk is full create a new chunk
+        if (chunkIsFull) {
+            //Add the message to the chunk
+            await this.createNewChunkAndAddMessage(message);
+            return;
+        }
+        //If the chunk is not full add the message to the chunk
+        await latestChunk.add(message);
+    }
+    public async fetch() {
+        const chunkIdentifier = await this.load();
+        //map the chunk identifier to chunk instances
+        const chunkInstances = chunkIdentifier.map(
+            (chunk: any) =>
+                new Chunk(this.db, this.enc, this.sizeLimit, chunk.id, []),
+        );
+
+        //update the conversation with the latest list of chunks
+        this.setLeafs(chunkInstances);
+    }
+    private async createNewChunkAndAddMessage(message: Envelop) {
+        //If the chunk is full create a new chunk
+        const emptyChunkIdentifer: ChunkIdentifier = {
+            //the id is the number of chunks
+            id: this.getLeafs().length,
+            //the timestamp of the first message
+            timestamp: message.message.metadata.timestamp,
+        };
+        //Create new chunk
+        const emptyChunk = await Chunk.createAndSafe(
+            this.db,
+            this.enc,
+            this.sizeLimit,
+            Chunk.computeKey(this.key, emptyChunkIdentifer),
+        );
+        //Add the chunk to the conversation
+        this.addLeaf(emptyChunk);
+        //finally the chunk that'll contain the message has been created and sits at the end of the list
+        await emptyChunk.add(message);
+        await this.save();
+    }
+}
+
+class Chunk extends Node {
```

Comment:
I think we need different Chunk types. Currently we only have chunks which can store messages but e.g. we also need a chunk that can store metadata related to a conversation (e.g. is the conversation archived).